From 37bd7feaabae73112ca8b4c3bdd03af258ec3f30 Mon Sep 17 00:00:00 2001
From: Aaradhana Sahu <quic_aarasahu@quicinc.com>
Date: Thu, 4 May 2023 21:26:31 +0530
Subject: [PATCH] ath12k: fix firmware assert during channel switch for
 multiple peer station

Currently, we are using deflink station instead of link_sta in
ath12k_mac_op_sta_rc_update function due to this firmware assert
is triggered when we perform channel switch for multiple peer
station in MLO.

Fix this issue by using link_sta in ath12k_mac_op_sta_rc_update.

Signed-off-by: Aaradhana Sahu <quic_aarasahu@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 22 ++++++++++++++++------
 1 file changed, 16 insertions(+), 6 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -7393,6 +7393,7 @@ static void ath12k_mac_op_sta_rc_update(
 	struct ath12k_link_sta *arsta;
 	struct ath12k_peer *peer;
 	u32 bw, smps;
+	struct ieee80211_link_sta *link_sta;
 
 	arvif = ahvif->link[link_id];
 	arsta = ahsta->link[link_id];
@@ -7427,10 +7428,18 @@ static void ath12k_mac_op_sta_rc_update(
 
 	spin_unlock_bh(&ar->ab->base_lock);
 
+	rcu_read_lock();
+	link_sta = rcu_dereference(sta->link[link_id]);
+
+	if (!link_sta) {
+		rcu_read_unlock();
+		return;
+	}
+
 	ath12k_dbg(ar->ab, ATH12K_DBG_MAC,
 		   "mac sta rc update for %pM changed %08x bw %d nss %d smps %d\n",
-		   arsta->addr, changed, sta->deflink.bandwidth, sta->deflink.rx_nss,
-		   sta->deflink.smps_mode);
+		   arsta->addr, changed, link_sta->bandwidth, link_sta->rx_nss,
+		   link_sta->smps_mode);
 
 	spin_lock_bh(&ar->data_lock);
 
@@ -7441,12 +7450,12 @@ static void ath12k_mac_op_sta_rc_update(
 	}
 
 	if (changed & IEEE80211_RC_NSS_CHANGED)
-		arsta->nss = sta->deflink.rx_nss;
+		arsta->nss = link_sta->rx_nss;
 
 	if (changed & IEEE80211_RC_SMPS_CHANGED) {
 		smps = WMI_PEER_SMPS_PS_NONE;
 
-		switch (sta->deflink.smps_mode) {
+		switch (link_sta->smps_mode) {
 		case IEEE80211_SMPS_AUTOMATIC:
 		case IEEE80211_SMPS_OFF:
 			smps = WMI_PEER_SMPS_PS_NONE;
@@ -7459,7 +7468,7 @@ static void ath12k_mac_op_sta_rc_update(
 			break;
 		default:
 			ath12k_warn(ar->ab, "Invalid smps %d in sta rc update for %pM\n",
-				    sta->deflink.smps_mode, arsta->addr);
+				    link_sta->smps_mode, arsta->addr);
 			smps = WMI_PEER_SMPS_PS_NONE;
 			break;
 		}
@@ -7470,6 +7479,7 @@ static void ath12k_mac_op_sta_rc_update(
 	arsta->changed |= changed;
 
 	spin_unlock_bh(&ar->data_lock);
+	rcu_read_unlock();
 
 	ieee80211_queue_work(hw, &arsta->update_wk);
 }
